blog.md

```diff
@@ -0,0 +1,203 @@
+However, this large number of parallel tasks also leads to another issue, which is the significant data payload. 
+Although the shared memory accessible by blocks is similar in speed to the L2 cache in CPUs, its capacity is limited. 
+Therefore, it is not possible to load all data into shared memory at once, and only the data required for the current block calculation can be loaded each time.
+Unfortunately, GEMM is a very memory-intensive operation, for example, when calculating the GEMM between a $M \times K$ matrix and a $K \times N$ matrix, if we use the naive way, i.e. evaluate the element in the result matrix one by one, we will have to load $M \times K \times N$ elements from the slow global memeory directly to registers in the whole process, and this generally far exceeds the data bandwidth of the GPU, resulting in severe performance issues.
+
+To avoid the heavy data loading, we first split the target matrix into blocks with size $BM \times BN$, and each GPU block will be used to calculate one of the tiled matrix, as shown in the fig below: 
+
+![Fig.1](figs/block.png)
+
+When calculating each block, we will only need to load matrices with size $BM \times BK$ and $BK \times BN$ for~$K / BK$ times from global memory to shared memory.
+In that case, the total data loading will be reduce to 
+$$
+    M \times N \times K \times \left( \frac{1}{BM} + \frac{1}{BN} \right)
+$$
+which is much smaller than the naive way.
+
+Then in each block, we further tile the matrix and use the registers to store the data, as shown by
+
+![Fig.1](figs/thread.png)
+
+The tiled target matrix will be further divided as small matrices with size $TM \times TN$, and each thread will be used to calculate one of the tiled matrix.
+During this process, the outer product way is used and data will be loaded from shared memory to registers, with the amount of $(TM + TN) \times BK$ in total.
+
+Then we replaced the operations directly.
+As we mentioned above, the `Tropical` type is `bitstype`, and the data storaged in memeories are simply the floating point numbers, which can be directly used by CUDA kernels.
+Although the tropical algebra can not use the fused multiple add core (FMA), we found that the operation add/mul and max/min can be done on FMA and ALU parallelly, which means that we can use the FMA to do the add/mul and ALU to do the max/min at the same time.
+Then after all the calculation is done, the target in the registers will be stored back to global memory directly.
+
+For the boundary elements, a padding strategy is used, we simply set the element which are not acctually in the matrix as the zero element of the corresponding algebra, so that they will not effect the result of the calculation.
+In our package, we set the parameters as
+$$
+    BM = 64,~BK = 32,~BN = 64,~TM = TN = 4.
```

Comment:
Maybe move this equation to the above?

Reply:
I moved that to above, after I finished the discussion about tiling.